Support valid DOIs ending in punctuation

mudge opened the pull request:
During some internal data clean-up work at @altmetric, we discovered DOIs that end with closing parentheses. These were being dropped by our extraction regular expression as we enforce a word boundary (which splits when a word character is followed by a non-word character such as a closing parenthesis).

By removing the strict word boundary, we can now pick up these closing parentheses and filter out any unwanted trailing punctuation in a second pass over the matches.

We do this by explicitly dropping the last character if it is punctuation but _not_ if the DOI ends in a single, balanced parenthetical group of characters or if it is an early Wiley DOI (which end in a certain sequence of punctuation).

This means that we can successfully extract valid DOIs such as 10.1002/0471221929.ch26(v) but not invalid ones such as 10.1007/978-3-319-12580-0).

mudge committed:
Support valid DOIs ending in punctuation

During some internal data clean-up work at @altmetric, we discovered
DOIs that end with closing parentheses. These were being dropped by our
extraction regular expression as we enforce a word boundary (which
splits when a word character is followed by a non-word character such as
a closing parenthesis).

By removing the strict word boundary, we can now pick up these closing
parentheses and filter out any unwanted trailing punctuation in a second
pass over the matches.

We do this by explicitly dropping the last character if it is
punctuation but _not_ if the DOI ends in a single, balanced
parenthetical group of characters or if it is an early Wiley DOI (which
end in a certain sequence of punctuation).

This means that we can successfully extract valid DOIs such as
10.1002/0471221929.ch26(v) but not invalid ones such as
10.1007/978-3-319-12580-0).

scottmatthewman commented on src/Doi.php:
```diff
@@ -3,10 +3,39 @@
+    private static function stripTrailingPunctuation($doi)
+    {
+        if (!preg_match('/[[:punct:]]$/', $doi) || preg_match('/\(.+\)$/', $doi) || preg_match(self::EARLY_WILEY_SUFFIX, $doi)) {
+            return strtolower($doi);
```
As we're using a method named `stripTrailingPunctuation` here, how would you feel about moving `strtolower()` out of the method? It feels like calling it within each `return` is indicating that it's actually a separate task from the punctuation stripping.

mudge replied:
🤔  I think I'd rather rename the method than loop twice over extracted matches.

mudge replied:
Having said that: I wonder if we could call `strtolower` once on the source string instead of on each individual match?

mudge committed:
Strip trailing Unicode punctuation from DOIs

Use PHP's UTF-8 mode for regular expressions and the punctuation
character property so that we can support DOIs with trailing, non-ASCII
punctuation (such as an ellipsis).

scottmatthewman commented on tests/DoiTest.php:
```diff
@@ -22,4 +32,37 @@ public function testLowercasesDois()
+    {
+        $this->assertEquals(
+            ['10.1002/(sici)1096-8644(199601)99:1<135::aid-ajpa8>3.0.co;2-#'],
+            Doi::extract('This is an example of an exotic DOI: 10.1002/(SICI)1096-8644(199601)99:1<135::AID-AJPA8>3.0.CO;2-#')
```
Would it be worthwhile to verify that the 'exotic' DOIs can also cope with trailing punctuation?

mudge replied:
This won't work for any of the special cases, e.g.:

```
1) Altmetric\Identifiers\DoiTest::testDiscardsTrailingPunctuationFromExoticDois
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '10.1002/(sici)1096-8644(199601)99:1<135::aid-ajpa8>3.0.co;2-#'
+    0 => '10.1002/(sici)1096-8644(199601)99:1<135::aid-ajpa8>3.0.co;2'
 )

2) Altmetric\Identifiers\DoiTest::testDiscardsTrailingPunctuationAfterBalancedParentheses
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '10.1130/2013.2502(04)'
+    0 => '10.1130/2013.2502(04'
 )
```

We'd have to add logic to support DOIs that both end in punctuation but also have superfluous trailing punctuation. Is it worth doing that here?

As it is, this improves our support for valid DOIs when passing, say, the contents of a `meta` tag but not when scanning full text as in a policy document.

mudge committed:
Remove trailing punctuation after Wiley DOIs

While our new logic supported valid DOIs ended in punctuation and
removed superfluous punctuation from otherwise valid DOIs, it didn't
support both at the same time: a valid, old Wiley DOI ending in #
couldn't be extracted if it also had trailing punctuation.

This fixes that by recursively stripping trailing punctuation one
character at a time until a valid DOI is found.

scottmatthewman commented on src/Doi.php:
```diff
@@ -3,10 +3,50 @@
+        }
 
-        return array_map('strtolower', $matches[0]);
+        return self::stripTrailingPunctuation(preg_replace('/\p{P}$/u', '', $doi));
```
😌 👍

mudge committed:
Don't extract DOIs with pure punctuation suffixes

As we now mutate DOIs after extracting them, it is possible for us to
invalidate the DOI during extraction. The simplest example of this is a
DOI with a suffix made entirely of trailing punctuation.

To fix this, ensure that we always revalidate a DOI when trying to strip
off any trailing punctuation.
